hal/mbed_lp_ticker_wrapper.cpp

```diff
@@ -0,0 +1,142 @@
+static mbed::Timeout *timeout;
+static uint64_t timeout_data[sizeof(mbed::Timeout) / 8];
+
+/**
+ * Initialize variables
+ */
+static void init_local()
+{
+    MBED_ASSERT(core_util_in_critical_section());
+
+    const ticker_info_t* info = lp_ticker_get_info();
+    if (info->bits >= 32) {
+        mask = 0xffffffff;
+    } else {
+        mask = ((uint64_t)1 << info->bits) - 1;
+    }
+
+    // Round us_per_tick up
+    timestamp_t us_per_tick = (1000000 + info->frequency - 1) / info->frequency;
+
+    min_delta = LOWPOWERTIMER_DELAY_TICKS;
+
+    // Add 1 tick to the min delta for the case where the clock transitions after you read it
+    // Add 4 microseconds to round up the micro second ticker time (which has a frequency of at least 250KHz - 4us period)
+    reschedule_us = (min_delta + 1) * us_per_tick + 4;
+
+    timeout = new (timeout_data) mbed::Timeout();
+}
+
+/**
+ * Call lp_ticker_set_interrupt with a value that is guaranteed to fire
+ *
+ * Assumptions
+ * -Only one low power clock tick can pass from the last read (last_read)
+ * -The closest an interrupt can fire is max_delta + 1
+ *
+ * @param last_read The last value read from lp_ticker_read
+ * @param timestamp The timestamp to trigger the interrupt at
+ */
+static void set_interrupt_safe(timestamp_t last_read, timestamp_t timestamp)
+{
+    MBED_ASSERT(core_util_in_critical_section());
+    uint32_t delta = (timestamp - last_read) & mask;
+    if (delta < min_delta + 2) {
+        timestamp = (last_read + min_delta + 2) & mask;
+    }
+    lp_ticker_set_interrupt(timestamp);
+}
+
+/**
+ * Set the low power ticker match time when hardware is ready
+ *
+ * This event is scheduled to set the lp timer after the previous write
+ * has taken effect and it is safe to write a new value without blocking.
+ * If the time has already passed then this function fires and interrupt
+ * immediately.
+ */
+static void set_interrupt_later()
+{
+    core_util_critical_section_enter();
+
+    timestamp_t current = lp_ticker_read();
+    if (_ticker_match_interval_passed(last_request, current, next)) {
+        lp_ticker_fire_interrupt();
+    } else {
+        set_interrupt_safe(current, next);
+        last_set_interrupt = lp_ticker_read();
+    }
+    timeout_pending = false;
+
+    core_util_critical_section_exit();
+}
+
+void lp_ticker_set_interrupt_wrapper(timestamp_t timestamp)
+{
+    core_util_critical_section_enter();
+
+    if (!init) {
+        init_local();
+        init = true;
+    }
+
+    timestamp_t current = lp_ticker_read();
+    if (pending) {
+        // Check if pending should be cleared
+        if (((current - last_set_interrupt) & mask) >= LOWPOWERTIMER_DELAY_TICKS) {
+            pending = false;
+        }
+    }
+
+    if (pending) {
+        next = timestamp;
+        last_request = current;
+        if (!timeout_pending) {
+            timeout->attach_us(set_interrupt_later, reschedule_us);
```

Comment:
I take it that this doesn't need to be unattached once the timeout elapses?

Reply:
Not needed. Detach only if you want to remove it early